cmd/otel-allocator/allocation/least_weighted.go

```diff
@@ -0,0 +1,200 @@
+		targetItemsCopy[k] = v
+	}
+	return targetItemsCopy
+}
+
+// Collectors returns a shallow copy of the collectors map.
+func (allocator *leastWeightedAllocator) Collectors() map[string]*Collector {
+	allocator.m.RLock()
+	defer allocator.m.RUnlock()
+	collectorsCopy := make(map[string]*Collector)
+	for k, v := range allocator.collectors {
+		collectorsCopy[k] = v
+	}
+	return collectorsCopy
+}
+
+// findNextCollector finds the next collector with fewer number of targets.
+// This method is called from within SetTargets and SetCollectors, whose caller
+// acquires the needed lock. This method assumes there are is at least 1 collector set
+// INVARIANT: allocator.collectors must have at least 1 collector set
+func (allocator *leastWeightedAllocator) findNextCollector() *Collector {
+	var col *Collector
+	for _, v := range allocator.collectors {
+		// If the initial collector is empty, set the initial collector to the first element of map
+		if col == nil {
+			col = v
+		} else if v.NumTargets < col.NumTargets {
+			col = v
+		}
+	}
+	return col
+}
+
+// addTargetToTargetItems assigns a target to the next available collector and adds it to the allocator's targetItems
+// This method is called from within SetTargets and SetCollectors, which acquire the needed lock.
+// This is only called after the collectors are cleared or when a new target has been found in the tempTargetMap
+// INVARIANT: allocator.collectors must have at least 1 collector set
+func (allocator *leastWeightedAllocator) addTargetToTargetItems(target *TargetItem) {
+	chosenCollector := allocator.findNextCollector()
+	targetItem := &TargetItem{
+		JobName:       target.JobName,
+		Link:          LinkJSON{Link: fmt.Sprintf("/jobs/%s/targets", url.QueryEscape(target.JobName))},
+		TargetURL:     target.TargetURL,
+		Label:         target.Label,
+		CollectorName: chosenCollector.Name,
+	}
+	allocator.targetItems[targetItem.Hash()] = targetItem
+	chosenCollector.NumTargets++
+	TargetsPerCollector.WithLabelValues(chosenCollector.Name, strategyName).Set(float64(chosenCollector.NumTargets))
+}
+
+func (allocator *leastWeightedAllocator) handleTargets(diff diff.Changes[*TargetItem]) {
+	// Check for removals
+	for k, target := range allocator.targetItems {
+		// if the current target is in the removals list
+		if _, ok := diff.Removals()[k]; ok {
+			c := allocator.collectors[target.CollectorName]
+			c.NumTargets--
+			delete(allocator.targetItems, k)
+			TargetsPerCollector.WithLabelValues(target.CollectorName, strategyName).Set(float64(c.NumTargets))
+		}
+	}
+
+	// Check for additions
+	for k, target := range diff.Additions() {
+		// Do nothing if the item is already there
+		if _, ok := allocator.targetItems[k]; ok {
+			continue
+		} else {
+			// Add target to target pool and assign a collector
+			allocator.addTargetToTargetItems(target)
+		}
+	}
+}
+
+func (allocator *leastWeightedAllocator) handleCollectors(diff diff.Changes[*Collector]) {
```

Comment:
Other than `NewAllocator`, this is the only function without a comment explaining what it does.  I think it's pretty clear what's happening but not sure if you want to add an explanation just for completeness.

Reply:
agreed, good idea to comment